### PR TITLE
adds saturating addition and subtraction operators

### PR DESCRIPTION
**Pull request by cc10512**
This PR introduces saturating arithmetic operators for signed and
unsigned integers of fixed width (per our discussion in the WG).

### CODE REVIEW

File: p4-16/spec/P4-16-spec.mdk [hunk 1]
```diff
@@ -2777,6 +2777,22 @@ keep the least-significant W bits of the value). In particular, P4
...
+result than a counter that overflows and wraps around. According to Wikipedia
+(https://en.wikipedia.org/wiki/Saturation_arithmetic) saturating arithmetic is
+as numerically close to the true answer as possible; for 8-bit binary signed
+arithmetic, when the correct answer is 130, it is considerably less surprising
+to get an answer of 127 from saturating arithmetic than to get an answer of
+−126 from modular arithmetic. Likewise, for 8-bit binary unsigned arithmetic,
+when the correct answer is 258, it is less surprising to get an answer of 255
+from saturating arithmetic than to get an answer of 2 from modular arithmetic.
+At this time, P4 defines saturating operations only for multiplication and
+addition. For an unsigned integer with bit-width of `W`, the minimum value is `0`
```
Comment (jafingerhut) [inline p4-16/spec/P4-16-spec.mdk]: Should "multiplication and addition" be replaced with "addition and subtraction"?  Those are the operators defined later.

Comment (vgurevich) [reply p4-16/spec/P4-16-spec.mdk]: +1 on replacing multiplication with subtraction.

File: p4-16/spec/P4-16-spec.mdk [hunk 2]
```diff
@@ -2815,6 +2831,8 @@ to bit-strings of the same width:
   by `|`.
 - Bitwise "complement" of a single bit-string, denoted by `~`.
 - Bitwise "xor" of two bit-strings of the same width, denoted by `^`.
+- Saturating addition, denoted by `|+|`.
```
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk]: Another possibility is `[+]`, which suggests confinement to a range.

File: p4-16/spec/P4-16-spec.mdk [hunk 3]
```diff
@@ -2852,6 +2870,10 @@ operations simply "wrap around", similar to C operations on unsigned values.
 Hence, attempting to represent large values using `W` bits will only keep
 the least-significant `W` bits of the value.
 
+P4 also supports saturating arithmetic (addition and subtraction) for signed
```
Comment (mbudiu-vmw) [inline p4-16/spec/P4-16-spec.mdk]: For signed integers I suspect we want this to work even when adding negative values.

File: p4-16/spec/P4-16-spec.mdk (hunk 2 above)
Comment (cc10512) [reply p4-16/spec/P4-16-spec.mdk]: I recommend using with `|+|`. `[]` are too overloaded.

File: p4-16/spec/P4-16-spec.mdk (hunk 1 above)
Comment (cc10512) [reply p4-16/spec/P4-16-spec.mdk]: fixed.

File: p4-16/spec/P4-16-spec.mdk (hunk 3 above)
Comment (cc10512) [reply p4-16/spec/P4-16-spec.mdk]: There is nothing in the text that suggests it should not. The minimum value is `-2^(W-1)`. Do you think we need to add additional clarification?

File: p4-16/spec/P4-16-spec.mdk [hunk 4]
```diff
@@ -2777,6 +2777,22 @@ keep the least-significant W bits of the value). In particular, P4
 does not have arithmetic exceptions---the result of an arithmetic
 operation is defined for all possible inputs.
 
+P4 target architectures may also support saturating arithmetic. All saturating
+operations are limited to a fixed range between a minimum and maximum value.
+Saturating arithmetic has advantages, in particular when used as counters. The
+the result of a saturating counter max-ing out is much closer to the real
+result than a counter that overflows and wraps around. According to Wikipedia
+(https://en.wikipedia.org/wiki/Saturation_arithmetic) saturating arithmetic is
```
Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk]: Convert to Markdown URL?

File: p4-16/spec/P4-16-spec.mdk [hunk 5]
```diff
@@ -2777,6 +2777,22 @@ keep the least-significant W bits of the value). In particular, P4
 does not have arithmetic exceptions---the result of an arithmetic
 operation is defined for all possible inputs.
 
+P4 target architectures may also support saturating arithmetic. All saturating
```
Comment (jnfoster) [inline p4-16/spec/P4-16-spec.mdk]: "may also" -> "may optionally"?

**Pull request merged**